File: co/system/backend/app/dependencies/battery.py
"""
Copyright (c) 2026 Clove Twilight
Licensed under the ESAL-1.3 Licence.
See LICENCE.md in the project root for full licence information.
"""

"""
Battery API authentication dependency
Simple header-based static key check, designed to be trivially callable from
Apple Shortcuts ("Get Contents of URL") and curl.
"""

from typing import Optional
from fastapi import Header, HTTPException, status

from app.services.battery_key_service import verify_battery_key


async def verify_battery_access(
    x_battery_key: Optional[str] = Header(None, alias="X-Battery-Key")
) -> bool:
    """
    Verify battery API access via the X-Battery-Key header.

    Raises:
        HTTPException: 401 if the header is missing or the key is invalid.
    """
    if not x_battery_key:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Missing X-Battery-Key header",
        )

    if not verify_battery_key(x_battery_key):
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid battery API key",
        )

    return True

File: system/backend/src/dependencies/battery.ts
/**
 * Copyright (c) 2026 Clove Twilight
 * Licensed under the ESAL-2.0 Licence.
 * See LICENCE.md in the project root for full licence information.
 */

/**
 * Battery API authentication dependency.
 * Simple header-based static key check, designed to be trivially callable
 * from Apple Shortcuts ("Get Contents of URL") and curl.
 *
 * NOTE: depends on `verifyBatteryKey` from ../services/battery_key_service.ts,
 * which hasn't been ported yet — this file won't compile until that exists.
 */

import type { Request, Response, NextFunction } from 'express';

import { HttpError } from '../core/errors.js';
import { verifyBatteryKey } from '../services/battery_key_service.js';

/**
 * Verify battery API access via the X-Battery-Key header.
 * On success, calls next(). On failure, forwards a 401 HttpError.
 */
export async function verifyBatteryAccess(req: Request, _res: Response, next: NextFunction): Promise<void> {
  const xBatteryKey = req.header('X-Battery-Key');

  if (!xBatteryKey) {
    next(new HttpError(401, 'Missing X-Battery-Key header'));
    return;
  }

  if (!(await verifyBatteryKey(xBatteryKey))) {
    next(new HttpError(401, 'Invalid battery API key'));
    return;
  }

  next();
}